### PR TITLE
Clotho Audio Captioning

### PR DESCRIPTION
**Pull request by Shikhar-S**
## What?

1. Add  clotho audio captioning recipe
2. Ads BEATs encoder to ESPnet
4. Add configs for BEATs encoder, BART decoder model
5. Add script to evaluate results using [FENSE](https://github.com/Labbeti/aac-metrics/tree/v0.5.4) metric.
6. Add data preparation scripts for audiocaps and clotho_chatgpt mixup as described in [this](https://dcase.community/documents/challenge2023/technical_reports/DCASE2023_Wu_31_t6a.pdf) paper.

Audio captioning Results
---

 cider_d             : 0.39208390185921266
 spice               : 0.1247477210504762
 spider              : 0.25841581145484444
 sbert_sim           : 0.5130076380936723
 fer                 : 0.03636363636363636
 fense               : 0.49523599610873387
 meteor              : 0.17313377768322902
 rouge_l             : 0.3479915684386986
 fer.add_tail_prob   : 0.04684687778353691
 fer.repeat_event_prob: 0.06736405938863754
 fer.repeat_adv_prob : 0.0016691883793100715
 fer.remove_conj_prob: 0.11576957255601883
 fer.remove_verb_prob: 0.19993385672569275
 fer.error_prob      : 0.3197185695171356
 spider_fl           : 0.24773266080817882

---

## Why?
Towards open-sourcing the winning implementation for [DCASE AAC challenge](https://dcase.community/challenge2024/task-automated-audio-captioning).

### CODE REVIEW

**Comment by Shikhar-S** [comment]
Still working towards finding correct hyper-parameters for the BEATs enc + BART dec model. Here is the detailed [report](https://api.wandb.ai/links/shikhar/ifdw7qbx) on current issues.

**Comment by Shikhar-S** [comment]
Layernorm bias and variance were getting re-initialized after pre-trained model loading in the earlier version. Fixed in this revision.

**Comment by Shikhar-S** [comment]
Updated README and paths to huggingface models. Please feel free to review now.

**Comment by sw005320** [comment]
@Shikhar-S, please fix the CI error
https://github.com/espnet/espnet/actions/runs/11830412428/job/32963877215?pr=5915

**Comment by sw005320** [comment]
@ftshijt, can you also review this PR?

File: egs2/clotho_v2/asr1/run_ft.sh [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+    --asr_tag ft.${ft_tag} \
+    --feats_normalize uttmvn \
+    --stage 1 \
+    --stop_stage 13 \
+    --asr_stats_dir ${expdir}/asr_stats_finetune \
+    --ngpu 2 \
+    --gpu_inference true \
+    --nj 8 \
+    --inference_nj 1 \
+    --max_wav_duration 30 \
```
Comment (ftshijt) [inline egs2/clotho_v2/asr1/run_ft.sh:38]: We can keep them as default in most of the case

File: egs2/clotho_v2/asr1/run_ft.sh [hunk 2]
```diff
@@ -0,0 +1,50 @@
+#!/usr/bin/env bash
+set -euo pipefail
+# Runs fine-tuning
+
+# wandb_init_args="--use_wandb true --wandb_project DCASE_AAC --wandb_model_log_interval 0"
+wandb_init_args=""
+other_args="$@"
+
+# expdir=/compute/babel-13-33/sbharad2/expdir
```
Comment (ftshijt) [inline egs2/clotho_v2/asr1/run_ft.sh:9]: Please remove the unused comments.

**Review by ftshijt — COMMENTED** [review]
Thanks for the update and your contribution! I left a few minor comments as follows:

File: egs2/clotho_v2/asr1/README.md [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+3. Launch with `run_pt.sh`
+
+For running fine-tuning:
+1. Set the value of `pt_tag` in `run_ft.sh` and
+2. Launch with `run_ft.sh`
+
+For running inference and scoring on all checkpoints `run_inference.sh` can be used by just changing the `name_tag`.
+
+## Trained checkpoints
+AAC Pre-trained model: https://huggingface.co/shikhar7ssu/dcase23.aac.pt
```
[the remark this answers is not in the document]
Comment (Shikhar-S) [reply egs2/clotho_v2/asr1/README.md]: Running stage 15 (upload to hf) with `hf_repo=espnet/DCASE23.AudioCaptioning.FineTuned` was not letting me upload due to authorization access. So I have manually uploaded the files to: https://huggingface.co/espnet/DCASE23.AudioCaptioning.PreTrained and https://huggingface.co/espnet/DCASE23.AudioCaptioning.FineTuned

File: egs2/clotho_v2/asr1/local/data.sh [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+    log "Fill the value of 'CLOTHO_V2' of db.sh"
+    exit 1
+fi
+
+log "stage 1: Data preparation"
+
+## DOWNLOAD DATA
+# If there is no argument, the default download directory is set to currentdir/downloads
+if [ $# -ne 1 ]; then
+    CLOTHO_V2_ROOT_DIR="$(pwd)/downloads"
```
[the remark this answers is not in the document]
Comment (Shikhar-S) [reply egs2/clotho_v2/asr1/local/data.sh]: Thanks for flagging. Updated.

File: egs2/clotho_v2/asr1/run_inference.sh [hunk 5]
```diff
@@ -0,0 +1,45 @@
+#!/usr/bin/env bash
```
[the remark this answers is not in the document]
Comment (Shikhar-S) [reply egs2/clotho_v2/asr1/run_inference.sh]: Thanks for the suggestion. I have removed unnecessary scripts.

File: egs2/TEMPLATE/asr1/db.sh [hunk 6]
```diff
@@ -220,6 +220,19 @@ SPRING_INX=downloads
...
+if [[ "$(hostname -d)" == delta.ncsa.illinois.edu ]]; then
+    AUDIOCAPS=/old_scratch/bbjs/shared/corpora/audiocaps/
+    CLOTHO_CHATGPT_MIXUP=/old_scratch/bbjs/slseanwu/dcase23_aac/beats_baseline/chatgpt/
+fi
+
+# For only CMU Babel environment on compute nodes
+if [[ "$(hostname)" == babel* ]]; then
+    AUDIOCAPS=/compute/babel-13-33/sbharad2/audiocaps/
+    CLOTHO_CHATGPT_MIXUP=/compute/babel-13-33/sbharad2/chatgpt/
+fi
```
Comment (sw005320) [inline egs2/TEMPLATE/asr1/db.sh:235]: `
AUDIOCAPS=
CLOTHO_CHATGPT_MIXUP=
```

File: egs2/clotho_v2/asr1/run_ft.sh (hunk 1 above)
Comment (Shikhar-S) [reply egs2/clotho_v2/asr1/run_ft.sh:26]: sorry, i do not fully understand this. what should i compare to? (also i need a different stats directory for fine-tuning)

File: egs2/clotho_v2/asr1/run_pt.sh [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+wandb_init_args=""
+other_args="$@"
+
+
+./asr.sh \
+    --asr_tag pt.${timestamp} \
+    --asr_speech_fold_length 1600 \
+    --feats_normalize uttmvn \
+    --stage 1 \
+    --stop_stage 13 \
```
Comment (sw005320) [inline egs2/clotho_v2/asr1/run_pt.sh:17]: why was it stopped at stage 13?

**Comment by sw005320** [comment]
Is it necessary to split the recipe with pre-training and fine-tuning?
As an "all-in-one" recipe, it is better to handle both in a single `run.sh`

File: egs2/clotho_v2/asr1/run_pt.sh (hunk 7 above)
Comment (Shikhar-S) [reply egs2/clotho_v2/asr1/run_pt.sh:17]: Willl update this to default values. 14 is model packing and 15 is uploading to HF (I did these separately as my setup wasn't letting me upload via terminal, currently idk why).

**Comment by Shikhar-S** [comment]
> Is it necessary to split the recipe with pre-training and fine-tuning? As an "all-in-one" recipe, it is better to handle both in a single `run.sh`

I see your point and it should be doable in a single script. I will update the scripts to make it work end to end, from data prep to printing final fine-tuned model's numbers.

File: egs2/clotho_v2/asr1/run_pt.sh (hunk 7 above)
Comment (sw005320) [reply egs2/clotho_v2/asr1/run_pt.sh:17]: model uploading does not happen in a default setup
https://github.com/espnet/espnet/blob/master/egs2/TEMPLATE/asr2/asr2.sh#L34
So, you do not need it anyway.

**Comment by Shikhar-S** [comment]
Closing this PR, added a better implementation in https://github.com/espnet/espnet/pull/5967
